Add bitmanip extension support

visitorckw opened the pull request:
Add support for the RISC-V bitmanip extensions Zba, Zbb, Zbc, and Zbs. These extensions optimize bitwise operations such as memory addressing, bitfield manipulation, counting leading/trailing zeros, and extracting subfields. All implementations have been validated through architectural tests, ensuring their correctness and reliability in emulating these operations.

visitorckw committed:
Enable testing for Zba, Zbb, Zbc, and Zbs extensions

Update the RISC-V architectural test to include support for testing the
Zba, Zbb, Zbc, and Zbs extensions.

With this change, users can run tests for these extensions using the
following command:

make arch-test RISCV_DEVICE=IMZbaZbbZbcZbs

visitorckw committed:
Add support for Zba address generation extension

Add support for the RISC-V Zba address generation extension. These
instructions combine shift and addition operations, enabling efficient
address calculation in pointer arithmetic and memory access patterns.

Supporting the Zba extension improves compatibility with the RISC-V
specification and enables optimized handling of address generation
tasks.

visitorckw committed:
Add support for Zbb basic bit-manipulation extension

Add support for the RISC-V Zbb basic bit-manipulation extension. This
extension introduces fundamental bitwise operations such as bit
counting, bit field extraction, and bit rotation, enabling efficient
data manipulation and low-level computation.

Supporting the Zbb extension improves compatibility with the RISC-V
specification and enables optimized handling of bitwise operations in a
wide range of applications.

visitorckw committed:
Add support for Zbc carry-less multiplication extension

Add support for the RISC-V Zbc carry-less multiplication extension.
This extension introduces instructions for carry-less multiplication,
enabling efficient operations in cryptographic algorithms,
error-correcting codes, and other specialized applications.

Supporting the Zbc extension improves compatibility with the RISC-V
specification and enables optimized handling of carry-less
multiplication tasks.

visitorckw committed:
Add support for Zbs single-bit instructions extension

Add support for the RISC-V Zbs single-bit instructions extension. This
extension introduces instructions for efficient manipulation of
individual bits, enabling optimized operations in bit-level processing
and control.

Supporting the Zbs extension improves compatibility with the RISC-V
specification and enhances performance in tasks requiring precise
single-bit operations.

visitorckw committed:
Add rv_ctz() for efficient count trailing zero operations

Introduce the rv_ctz() function to compute the count of trailing zeros
in a 32-bit integer. This function serves as a prerequisite for
implementing the ctz instruction in the RISC-V Zbb extension.

visitorckw committed:
Add rv_popcount() for efficient population count operations

Introduce the rv_popcount() function to compute the population count
(number of set bits) in a 32-bit integer. This function is a
prerequisite for implementing the cpop instruction in the RISC-V Zbb
extension.

visitorckw committed:
CI: Add Zba, Zbb, Zbc, and Zbs tests to CI pipeline

Add tests for the Zba, Zbb, Zbc, and Zbs extensions to the CI pipeline,
ensuring that these extensions are validated automatically during
continuous integration.

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 73dea0c5e3a9473b523b0e5536c061d6b3521d19 | Previous: e7ded18da549393a3e2c96550ac4fadf51bd06ab | Ratio |
|-|-|-|-|
| `Dhrystone` | `1337` Average DMIPS over 10 runs | `1301` Average DMIPS over 10 runs | `0.97` |
| `Coremark` | `975.043` Average iterations/sec over 10 runs | `977.65` Average iterations/sec over 10 runs | `1.00` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv reviewed (changes requested):
Update the top-level `README.md` as well.

visitorckw committed:
Mention Zba, Zbb, Zbc, and Zbs extensions in README

Update the README to include details about the RISC-V Zba, Zbb, Zbc,
and Zbs bit manipulation extensions.

ChinYikMing commented on src/decode.c:
```diff
@@ -478,6 +514,26 @@ static inline bool op_op_imm(rv_insn_t *ir, const uint32_t insn)
         break;
     case 5:
+#if RV32_HAS(Zbb)
+        if ((ir->imm >> 5) == 0b0110000) { /* rori */
```
Remove bracket for consistency.

jserv commented:
Thank @visitorckw for contributing!
